Make -fclash-float-support the default and deprecate the flag

If you have a project that legitimately uses `Float` and forget to include the command-line parameter `-fclash-float-support`, the generated HDL will silently omit anything that works with `Float`s. It would be much better if Clash at least warned that something seems to go wrong.

Using the regular operators, you currently get this (with `-fclash-float-support --vhdl`):
```
PlainFloat.hs:9:1: error:
    
    Clash.Netlist.BlackBox(311): No blackbox found for: GHC.Prim.plusFloat#. Did you forget to include directories containing primitives? You can use '-i/my/prim/dir' to achieve this.
    
    CallStack (from HasCallStack):
      extractPrimWarnOrFail, called at src/Clash/Netlist/Util.hs:841:25 in clash-lib-1.5.0-inplace:Clash.Netlist.Util
    
    
    The source location of the error is not exact, only indicative, as it is acquired 
    after optimizations. The actual location of the error can be in a function that is 
    inlined. To prevent inlining of those functions, annotate them with a NOINLINE pragma.
  |
9 | topEntity = (+)
  | ^^^^^^^^^
```

The actual float support for Xilinx is indeed with explicitly used functions.

--- a/Clash.hs
+++ b/Clash.hs
@@ -63,7 +63,7 @@ doHDL b src = do
   putStrLn $ "Loading dependencies took " ++ prepStartDiff
 
   generateHDL (buildCustomReprs reprs) domainConfs bindingsMap (Just b) primMap tcm tupTcm
-    (ghcTypeToHWType WORD_SIZE_IN_BITS True) ghcEvaluator evaluator topEntities Nothing
+    (ghcTypeToHWType WORD_SIZE_IN_BITS) ghcEvaluator evaluator topEntities Nothing
     defClashOpts{opt_cachehdl = False, opt_debug = debugSilent, opt_clear = True}
     startTime
 

--- a/benchmark/common/BenchmarkCommon.hs
+++ b/benchmark/common/BenchmarkCommon.hs
@@ -51,7 +51,7 @@ defaultTests =
 
 typeTrans :: (CustomReprs -> TyConMap -> Type ->
               State HWMap (Maybe (Either String FilteredHWType)))
-typeTrans = ghcTypeToHWType WORD_SIZE_IN_BITS True
+typeTrans = ghcTypeToHWType WORD_SIZE_IN_BITS
 
 opts :: [FilePath] -> ClashOpts
 opts idirs =
@@ -59,7 +59,6 @@ opts idirs =
       opt_cachehdl=False
     , opt_clear=True
     , opt_errorExtra = True
-    , opt_floatSupport = True
     , opt_importPaths=idirs
     , opt_specLimit=100 -- For "ManyEntitiesVaried"
     }

--- a/changelog/2022-01-23T09_15_21+00_00_remove_ffloat_support
+++ b/changelog/2022-01-23T09_15_21+00_00_remove_ffloat_support
@@ -1,0 +1,1 @@
+DEPRECATED: Deprecated '-fclash-float-support': it is now on by default and can't be turned off.

--- a/clash-ghc/src-bin-8.10/Clash/GHCi/UI.hs
+++ b/clash-ghc/src-bin-8.10/Clash/GHCi/UI.hs
@@ -2195,7 +2195,6 @@ makeHDL backend startAction optsRef srcs = do
               opts0  <- readIORef optsRef
               let opts1  = opts0 { opt_color = useColor dflags }
               let iw     = opt_intWidth opts1
-                  fp     = opt_floatSupport opts1
                   syn    = opt_hdlSyn opts1
                   color  = opt_color opts1
                   esc    = opt_escapedIds opts1
@@ -2243,7 +2242,7 @@ makeHDL backend startAction optsRef srcs = do
                   primMap
                   tcm
                   tupTcm
-                  (ghcTypeToHWType iw fp)
+                  (ghcTypeToHWType iw)
                   ghcEvaluator
                   evaluator
                   topEntities

--- a/clash-ghc/src-bin-861/Clash/GHCi/UI.hs
+++ b/clash-ghc/src-bin-861/Clash/GHCi/UI.hs
@@ -2026,7 +2026,6 @@ makeHDL backend startAction optsRef srcs = do
               opts0 <- readIORef optsRef
               let opts1  = opts0 { opt_color = useColor dflags }
                   iw     = opt_intWidth opts1
-                  fp     = opt_floatSupport opts1
                   syn    = opt_hdlSyn opts1
                   color  = opt_color opts1
                   esc    = opt_escapedIds opts1
@@ -2073,7 +2072,7 @@ makeHDL backend startAction optsRef srcs = do
                   primMap
                   tcm
                   tupTcm
-                  (ghcTypeToHWType iw fp)
+                  (ghcTypeToHWType iw)
                   ghcEvaluator
                   evaluator
                   topEntities

--- a/clash-ghc/src-bin-881/Clash/GHCi/UI.hs
+++ b/clash-ghc/src-bin-881/Clash/GHCi/UI.hs
@@ -2117,7 +2117,6 @@ makeHDL backend startAction optsRef srcs = do
               opts0  <- readIORef optsRef
               let opts1  = opts0 { opt_color = useColor dflags }
               let iw     = opt_intWidth opts1
-                  fp     = opt_floatSupport opts1
                   syn    = opt_hdlSyn opts1
                   color  = opt_color opts1
                   esc    = opt_escapedIds opts1
@@ -2165,7 +2164,7 @@ makeHDL backend startAction optsRef srcs = do
                   primMap
                   tcm
                   tupTcm
-                  (ghcTypeToHWType iw fp)
+                  (ghcTypeToHWType iw)
                   ghcEvaluator
                   evaluator
                   topEntities

--- a/clash-ghc/src-bin-9.0/Clash/GHCi/UI.hs
+++ b/clash-ghc/src-bin-9.0/Clash/GHCi/UI.hs
@@ -2234,7 +2234,6 @@ makeHDL backend startAction optsRef srcs = do
               opts0  <- readIORef optsRef
               let opts1  = opts0 { opt_color = useColor dflags }
               let iw     = opt_intWidth opts1
-                  fp     = opt_floatSupport opts1
                   syn    = opt_hdlSyn opts1
                   color  = opt_color opts1
                   esc    = opt_escapedIds opts1
@@ -2282,7 +2281,7 @@ makeHDL backend startAction optsRef srcs = do
                   primMap
                   tcm
                   tupTcm
-                  (ghcTypeToHWType iw fp)
+                  (ghcTypeToHWType iw)
                   ghcEvaluator
                   evaluator
                   topEntities

--- a/clash-ghc/src-ghc/Clash/GHC/ClashFlags.hs
+++ b/clash-ghc/src-ghc/Clash/GHC/ClashFlags.hs
@@ -77,7 +77,7 @@ flagsClash r = [
   , defFlag "fclash-evaluator-fuel-limit"        $ IntSuffix (liftEwM . setEvaluatorFuelLimit r)
   , defFlag "fclash-intwidth"                    $ IntSuffix (setIntWidth r)
   , defFlag "fclash-error-extra"                 $ NoArg (liftEwM (setErrorExtra r))
-  , defFlag "fclash-float-support"               $ NoArg (liftEwM (setFloatSupport r))
+  , defFlag "fclash-float-support"               $ NoArg (setFloatSupport r)
   , defFlag "fclash-component-prefix"            $ SepArg (liftEwM . setComponentPrefix r)
   , defFlag "fclash-old-inline-strategy"         $ NoArg (liftEwM (setOldInlineStrategy r))
   , defFlag "fclash-no-escaped-identifiers"      $ NoArg (liftEwM (setNoEscapedIds r))
@@ -271,8 +271,9 @@ setHdlSyn r s = case readMaybe s of
 setErrorExtra :: IORef ClashOpts -> IO ()
 setErrorExtra r = modifyIORef r (\c -> c {opt_errorExtra = True})
 
-setFloatSupport :: IORef ClashOpts -> IO ()
-setFloatSupport r = modifyIORef r (\c -> c {opt_floatSupport = True})
+setFloatSupport :: IORef ClashOpts -> EwM IO ()
+setFloatSupport _ =
+  addWarn "Deprecated flag: -fclash-float-support is always enabled from Clash 1.6 and onwards"
 
 setComponentPrefix
   :: IORef ClashOpts

--- a/clash-ghc/src-ghc/Clash/GHC/NetlistTypes.hs
+++ b/clash-ghc/src-ghc/Clash/GHC/NetlistTypes.hs
@@ -45,8 +45,6 @@ ghcTypeToHWType
   :: Int
   -- ^ Integer width. The width Clash assumes an Integer to be (instead of it
   -- begin an arbitrarily large, runtime sized construct).
-  -> Bool
-  -- ^ Float support
   -> CustomReprs
   -- ^ Custom bit representations
   -> TyConMap
@@ -54,7 +52,7 @@ ghcTypeToHWType
   -> Type
   -- ^ Type to convert to HWType
   -> State HWMap (Maybe (Either String FilteredHWType))
-ghcTypeToHWType iw floatSupport = go
+ghcTypeToHWType iw = go
   where
     -- returnN :: HWType ->
     returnN t = return (FilteredHWType t [])
@@ -112,14 +110,14 @@ ghcTypeToHWType iw floatSupport = go
         "GHC.Prim.Word#"                -> returnN (Unsigned iw)
         "GHC.Prim.Int64#"               -> returnN (Signed 64)
         "GHC.Prim.Word64#"              -> returnN (Unsigned 64)
-        "GHC.Prim.Float#" | floatSupport -> returnN (BitVector 32)
-        "GHC.Prim.Double#" | floatSupport -> returnN (BitVector 64)
+        "GHC.Prim.Float#"               -> returnN (BitVector 32)
+        "GHC.Prim.Double#"              -> returnN (BitVector 64)
         "GHC.Prim.ByteArray#"           ->
           throwE $ "Can't translate type: " ++ showPpr ty
 
         "GHC.Types.Bool"                -> returnN Bool
-        "GHC.Types.Float" | floatSupport-> returnN (BitVector 32)
-        "GHC.Types.Double" | floatSupport -> returnN (BitVector 64)
+        "GHC.Types.Float"               -> returnN (BitVector 32)
+        "GHC.Types.Double"              -> returnN (BitVector 64)
         "GHC.Prim.~#"                   -> returnN (Void Nothing)
 
         "Clash.Signal.Internal.Signal" ->

--- a/clash-lib/src/Clash/Driver/Manifest.hs
+++ b/clash-lib/src/Clash/Driver/Manifest.hs
@@ -129,12 +129,11 @@ data Manifest
   = Manifest
   { manifestHash :: ByteString
     -- ^ Hash digest of the TopEntity and all its dependencies.
-  , successFlags  :: (Int,Int,Bool)
+  , successFlags  :: (Int, Int)
     -- ^ Compiler flags used to achieve successful compilation:
     --
     --   * opt_inlineLimit
     --   * opt_specLimit
-    --   * opt_floatSupport
   , ports :: [ManifestPort]
     -- ^ Ports in the generated @TopEntity@.
   , componentNames :: [Text]
@@ -309,7 +308,7 @@ mkManifest backend domains ClashOpts{..} Component{..} components deps files top
   , componentNames = map Id.toText compNames
   , topComponent = Id.toText componentName
   , fileNames = files
-  , successFlags = (opt_inlineLimit, opt_specLimit, opt_floatSupport)
+  , successFlags = (opt_inlineLimit, opt_specLimit)
   , domains = domains
   , transitiveDependencies = map (nameOcc . varName) deps
   }
@@ -410,12 +409,9 @@ readFreshManifest tops (bindingsMap, topId) primMap opts@(ClashOpts{..}) clashMo
       -- they could affect successful compilation, and use that information
       -- to decide whether to use caching or not (see: XXXX).
       --
-      -- 1. termination measures
+      -- 5. termination measures
     , opt_inlineLimit = 20
     , opt_specLimit = 20
-
-      -- 2. Float support
-    , opt_floatSupport = False
 
       -- Finally, also ignore the HDL dir setting, because when a user moves the
       -- entire dir with generated HDL, they probably still want to use that as
@@ -434,15 +430,11 @@ readFreshManifest tops (bindingsMap, topId) primMap opts@(ClashOpts{..}) clashMo
     )
 
   checkManifest manifest@Manifest{manifestHash,successFlags}
-    | (cachedInline, cachedSpec, cachedFloat) <- successFlags
+    | (cachedInline, cachedSpec) <- successFlags
 
     -- Higher limits shouldn't affect HDL
     , cachedInline <= opt_inlineLimit
     , cachedSpec <= opt_specLimit
-
-    -- /Enabling/ float support should compile more designs. Of course, keeping
-    -- the same value for float support shouldn't invalidate caches either.
-    , ((cachedFloat && not opt_floatSupport) || (cachedFloat == opt_floatSupport))
 
     -- Callgraph hashes should correspond
     , manifestHash == topHash

--- a/clash-lib/src/Clash/Driver/Types.hs
+++ b/clash-lib/src/Clash/Driver/Types.hs
@@ -315,9 +315,6 @@ data ClashOpts = ClashOpts
   -- ^ Synthesis target. See "HdlSyn" for available options.
   , opt_errorExtra :: Bool
   -- ^ Show additional information in error messages
-  , opt_floatSupport :: Bool
-  -- ^ Treat floats as a BitVector. Note that operations on floating are still
-  -- not supported, use vendor primitives instead.
   , opt_importPaths :: [FilePath]
   -- ^ Paths where Clash should look for modules
   , opt_componentPrefix :: Maybe Text
@@ -381,7 +378,6 @@ instance Eq ClashOpts where
     opt_hdlDir s0 == opt_hdlDir s1 &&
     opt_hdlSyn s0 == opt_hdlSyn s1 &&
     opt_errorExtra s0 == opt_errorExtra s1 &&
-    opt_floatSupport s0 == opt_floatSupport s1 &&
     opt_importPaths s0 == opt_importPaths s1 &&
     opt_componentPrefix s0 == opt_componentPrefix s1 &&
     opt_newInlineStrat s0 == opt_newInlineStrat s1 &&
@@ -419,7 +415,6 @@ instance Hashable ClashOpts where
     opt_hdlDir `hashWithSalt`
     opt_hdlSyn `hashWithSalt`
     opt_errorExtra `hashWithSalt`
-    opt_floatSupport `hashWithSalt`
     opt_importPaths `hashWithSalt`
     opt_componentPrefix `hashWithSalt`
     opt_newInlineStrat `hashWithSalt`
@@ -457,7 +452,6 @@ defClashOpts
   , opt_hdlDir              = Nothing
   , opt_hdlSyn              = Other
   , opt_errorExtra          = False
-  , opt_floatSupport        = False
   , opt_importPaths         = []
   , opt_componentPrefix     = Nothing
   , opt_newInlineStrat      = True

--- a/tests/src/Test/Tasty/Clash/CoreTest.hs
+++ b/tests/src/Test/Tasty/Clash/CoreTest.hs
@@ -49,7 +49,6 @@ mkClashOpts :: ClashOpts
 mkClashOpts = defClashOpts
   { opt_cachehdl     = False
   , opt_errorExtra   = True
-  , opt_floatSupport = True
   }
 
 mkBackend

--- a/tests/src/Test/Tasty/Clash/NetlistTest.hs
+++ b/tests/src/Test/Tasty/Clash/NetlistTest.hs
@@ -65,13 +65,12 @@ typeTrans
   -> TyConMap
   -> Type
   -> State HWMap (Maybe (Either String FilteredHWType))
-typeTrans = ghcTypeToHWType WORD_SIZE_IN_BITS True
+typeTrans = ghcTypeToHWType WORD_SIZE_IN_BITS
 
 mkClashOpts :: ClashOpts
 mkClashOpts = defClashOpts
   { opt_cachehdl     = False
   , opt_errorExtra   = True
-  , opt_floatSupport = True
   }
 
 type family TargetToState (target :: HDL) where

